compiler/grainlsp/utils.re

```diff
@@ -0,0 +1,814 @@
+
+    switch (args) {
+    | [] => ""
+    | _ => "(" ++ comma_join(~sep=", ", ~print_fn=print_sig, args) ++ ")"
+    }
+  | TTyVar(v) =>
+    // log("TTyVar");
+    switch (v) {
+    | None => ""
+    | Some(txt) => txt
+    }
+  | TTyUniVar(v) =>
+    //log("TTyUniVar");
+    switch (v) {
+    | None => ""
+    | Some(var) => var
+    }
+  | TTyPoly(te, _) => lens_sig(~log, ~depth=depth + 1, ~env, te)
+  | TTyLink(te) =>
+    //   log("TTyLink");
+    lens_sig(~log, ~depth=depth + 1, ~env, te)
+
+  | TTySubst(t) =>
+    //   log("TTySubst");
+    lens_sig(~log, ~depth=depth + 1, ~env, t)
+  | TTyConstr(path, types, r) =>
+    //  log("TTyConstr");
+
+    let decl = Env.find_type(path, env);
+
+    let tk = decl.type_kind;
+
+    let type_text =
+      switch (types) {
+      | [] => ""
+      | _ => "<" ++ comma_join(~sep=", ", ~print_fn=print_sig, types) ++ ">"
+      };
+
+    switch (tk) {
+    | TDataRecord(fields) =>
+      let labelText =
+        List.fold_left(
+          (acc, field: Types.record_field) => {
+            let existing =
+              if (acc == "") {
+                acc;
+              } else {
+                acc ++ ",\n";
+              };
+            let typeInf = lens_sig(~log, ~env, field.rf_type);
+            let rf_name = field.rf_name;
+
+            existing ++ "  " ++ rf_name.name ++ ": " ++ typeInf;
+          },
+          "",
+          fields,
+        );
+
+      "record "
+      ++ print_path(path)
+      ++ type_text
+      ++ " {\n"
+      ++ labelText
+      ++ "\n}";
+    | TDataVariant(decls) =>
+      //  log("TDataVariant");
+      let declText =
+        List.fold_left(
+          (acc, decl: Types.constructor_declaration) => {
+            let existing =
+              if (acc == "") {
+                acc;
+              } else {
+                acc ++ ",\n";
+              };
+            let decl_name = decl.cd_id.name;
+
+            let args = decl.cd_args;
+
+            let decl_contructions =
+              switch (args) {
+              | TConstrTuple(types) =>
+                switch (types) {
+                | [] => ""
+                | _ =>
+                  " ("
+                  ++ comma_join(~sep=",\n", ~print_fn=print_sig, types)
+                  ++ ")"
+                }
+              | TConstrSingleton => ""
+              };
+
+            existing ++ "  " ++ decl_name ++ decl_contructions;
+          },
+          "",
+          decls,
+        );
+
+      let parts =
+        switch (path) {
+        | PIdent(ident) => ("", ident.name)
+        | PExternal(mod_path, name, _) => (
+            switch (mod_path) {
+            | PIdent(ident) => ident.name
+            | PExternal(mod_path, name, _) => ""
+            },
+            name,
+          )
+        };
+
+      let (modname, typ) = parts;
+
+      if (modname == "Pervasives") {
+        typ ++ type_text;
+      } else if (typ == "Void") {
+        "Void";
+      } else {
+        "enum "
```

Comment:
Should we be using oprint or something here?

Reply:
Not sure how to, need some help I think.

Reply:
Jinx

Reply:
Yes, no longer needed.